domehunter/dome_control.py

```diff
@@ -355,81 +386,176 @@ def calibrate_dome_encoder_counts(self, num_cal_rotations=2):
         self._encoder_count = self._az_to_ticks(Longitude(az * u.deg))
         return
 
 ###############################################################################
 # Private Methods
 ###############################################################################
 
+    def _thread_condition(self, trigger_condition):
+        """Condition monitoring thread for movement commands.
+
+        Will return when:
+            - trigger_condition is true
+            - abort event is triggered
+            - thread running time exceeds self.wait_timeout
+
+        """
+        goingtoaz = False
+        if trigger_condition.__name__ == '_goto_az_condition':
+            goingtoaz = True
+        start = time.time()
+        while True:
+            time.sleep(0.1)
+            wait_time = time.time() - start
+            if trigger_condition():
+                logger.debug(f'Monitoring thread triggered by {trigger_condition.__name__}')
+                break
+            elif self._abort_event.is_set():
+                logger.debug(f'Monitoring thread triggered by abort event')
+                break
+            elif wait_time > self.wait_timeout:
+                logger.debug(f'Monitoring thread triggered by timeout [{self.wait_timeout}s]')
+                break
+            elif goingtoaz and self.testing is True:
+                # if testing simulate a tick for every cycle of while loop
+                self._simulate_ticks(num_ticks=1)
+        logger.debug('Monitor Thread: Stopping dome movement')
+        self._stop_moving()
+        # reset various dome state variables/events
+        self._move_event.clear()
+        self._calibrating = False
+        self.rotation_count = 0
+        return
+
+    def _goto_az_condition(self):
+        """Determines if current azimuth is within tolerance of target azimuth.
+
+        Returns
+        -------
+        bool
+            Returns true if self.dome_az is within tolerance of target azimuth.
+
+        """
+        delta_az = self.current_direction * (self.target_az - self.dome_az).wrap_at('180d')
+        logger.debug(f'Delta_az is {delta_az}')
+        return delta_az <= self.az_position_tolerance
+
+    def _find_home_condition(self):
+        """Return True if the dome is at home."""
+        # need a callable method for _thread_condition
+        return self._home_sensor.is_active
+
+    def _calibration_completed(self):
+        """Return True if both calibrating and move_event flags are True"""
+        cal_in_progress = self._calibrating and self._move_event.is_set()
+        return not cal_in_progress
+
+    def _calibrate(self):
+        """Calibration method to be called from within a non-blocking thread"""
+        start = time.time()
+        # if self.testing:
+        #     for i in range(self.num_cal_rotations):
+        #         pass
+        first_rotation = threading.Thread(target=self._simulate_rotation)
+        first_rotation.start()
```

Comment:
Do you need these? I don't see them anywhere else and this is calling a thread from within a thread.

Reply:
So there is the calibration monitoring thread, that is checking for the right number of calibration rotations, abort events and timeouts. This needs to be making those checks regularly. For testing mode I also need to simulate the rotations, so they need to be run in a non-blocking thread.
So to begin I start an initial rotation before starting the `while` loop and include a clause within that loop that checks if the previous simulated rotation thread has completed. If the number of desired is rotations has been met the monitor thread will terminate, otherwise it will start another simulated rotation in its own thread.
Maybe there is a better way of doing this but the idea was that the thread monitoring calibration needs run separately to the thread simulating the rotation. I could have set the rotation threads up with a delay from within the main `calibrate_dome_encoder_counts()` method but you'd have to specify the timing of when each simulated rotation would start. I just thought that seemed messy/could lead to over lapping threads (a new rotation thread starts before the last is done etc).